asf_heat_pump_suitability/pipeline/prepare_features/grid_capacity.py

```diff
@@ -0,0 +1,758 @@
+    ).clip(upper=100, lower=0)
+
+    # Calculate excess or deficit capacity
+    lsoa_data["capacity_difference_kw"] = lsoa_data["headroom_kw"] - (
+        lsoa_data["household_count"] * power_per_heatpump
+    )
+
+    return lsoa_data
+
+
+def calculate_grid_capacity() -> pd.DataFrame:
+    """
+    Calculate the grid capacity for heat pump installations across all LSOAs.
+
+    This function performs the following steps:
+    1. Generate and process substation data from all grid operators
+    2. Load and process LSOA and household data
+    3. Distribute substation headroom to LSOAs
+    4. Assess heat pump suitability based on available capacity
+
+    Returns:
+        DataFrame with grid capacity assessment results for each LSOA
+    """
+    # Generate and process substation data
+    substations = generate_substations_df()
+    substations["firm_capacity_mva"] = substations["firm_capacity_mva"].apply(
+        _parse_capacity
+    )
+    substations["peak_demand_mva"] = substations["peak_demand_mva"].apply(
+        _parse_capacity
+    )
+
+    # Apply substation rating scaling factor
+    substations["firm_capacity_mva"] = (
+        substations["firm_capacity_mva"] * SUBSTATION_SCALING_FACTOR
+    )
+
+    # Calculate headroom per substation
+    substations["headroom_mva"] = (
+        substations["firm_capacity_mva"] - substations["peak_demand_mva"]
+    )
+
+    # Load and process LSOA boundary data
+    lsoa_boundaries = load_s3_data(
+        "asf-heat-pump-suitability",
+        "source_data/Output_Areas_2021_EW_BGC_V2_4299916833741807639.geojson",
+    )
+    lsoa_gdf = aggregate_oa_to_lsoa(lsoa_boundaries)
+
+    # Load and process household data
+    households_df = get_datasets.get_df_ons_number_of_households()
+    households_df = households_df.rename(
+        mapping={"mnemonic": "LSOA21CD", "2021": "household_count"}
+    ).to_pandas()
+
+    # Process substation and LSOA data
+    headroom_df = process_substation_lsoa_data(substations, lsoa_gdf, households_df)
+
+    # Assess heat pump suitability
+    consumption_per_heatpump = POWER_PER_HEATPUMP / COEFFICIENT_OF_PERFORMANCE
+    result = assess_heatpump_suitability(
+        headroom_df, consumption_per_heatpump, TAKEUP_RATE
+    )
+
+    # Rename LSOA column for consistency
+    result = result.rename(columns={"LSOA21CD": "lsoa21"})
+
+    return result
+
+
+# Example usage
+if __name__ == "__main__":
+    grid_capacity_results = calculate_grid_capacity()
+    print(grid_capacity_results.head())
+    print(f"Total LSOAs assessed: {len(grid_capacity_results)}")
+    print(
+        f"LSOAs with sufficient grid capacity: {grid_capacity_results['has_grid_capacity'].sum()}"
+    )
+    print(
+        f"Percentage of LSOAs with sufficient capacity: {(grid_capacity_results['has_grid_capacity'].mean() * 100):.2f}%"
+    )
+
+    # Optionally, save results to a file
+    # grid_capacity_results.to_csv("grid_capacity_assessment_results.csv", index=False)
```

Comment:
you can put this as an arg to the script. here's my arg parser you can use as a baseline if you wanted to add it to save you the time:

```
def parse_arguments() -> argparse.Namespace:
    """
    Create ArgumentParser and parse.

    Returns:
        argparse.Namespace: populated `Namespace`
    """
    parser = ArgumentParser()
    parser.add_argument(
        "--save_as",
        help="Path to save grid capacity results to. If unspecified, save with default filename.",
        type=str,
        required=False,
        default=None,
    )

    return parser.parse_args()
```
Then in script you can set the default name.